tensorzero-internal/src/endpoints/datasets.rs

```diff
@@ -0,0 +1,305 @@
+        }));
+    }
+    let demonstration: Demonstration = serde_json::from_str(&result).map_err(|e| {
+        Error::new(ErrorDetails::Serialization {
+            message: format!(
+                "Failed to deserialize demonstration ClickHouse response: {}",
+                e
+            ),
+        })
+    })?;
+    Ok(demonstration)
+}
+
+#[derive(Deserialize)]
+#[serde(deny_unknown_fields, tag = "function_type", rename_all = "snake_case")]
+pub enum TaggedInferenceDatabaseInsert {
+    Chat(ChatInferenceDatabaseInsert),
+    Json(JsonInferenceDatabaseInsert),
+}
+
+async fn query_inference_for_datapoint(
+    clickhouse: &ClickHouseConnectionInfo,
+    inference_id: Uuid,
+) -> Result<TaggedInferenceDatabaseInsert, Error> {
+    let result: String = clickhouse
+        .run_query(
+            r#"
+            SELECT
+  uint_to_uuid(i.id_uint) AS id,
+
+  -- Common columns (pick via IF)
+  IF(i.function_type = 'chat', c.function_name, j.function_name) AS function_name,
+  IF(i.function_type = 'chat', c.variant_name,   j.variant_name)   AS variant_name,
+  IF(i.function_type = 'chat', c.episode_id,     j.episode_id)     AS episode_id,
+  IF(i.function_type = 'chat', c.input,          j.input)          AS input,
+  IF(i.function_type = 'chat', c.output,         j.output)         AS output,
+
+  -- Chat-specific columns
+  IF(i.function_type = 'chat', c.tool_params, '') AS tool_params,
+
+  -- Inference params (common name in the union)
+  IF(i.function_type = 'chat', c.inference_params, j.inference_params) AS inference_params,
+
+  -- Processing time
+  IF(i.function_type = 'chat', c.processing_time_ms, j.processing_time_ms) AS processing_time_ms,
+
+  -- JSON-specific column
+  IF(i.function_type = 'json', j.output_schema, '') AS output_schema,
+
+  -- Timestamps & tags
+  IF(i.function_type = 'chat',
+   formatDateTime(c.timestamp, '%Y-%m-%dT%H:%i:%SZ'),
+   formatDateTime(j.timestamp, '%Y-%m-%dT%H:%i:%SZ')
+) AS timestamp,
+  IF(i.function_type = 'chat', c.tags,      j.tags)      AS tags,
+
+  -- Discriminator itself
+  i.function_type
+
+FROM InferenceById i
+LEFT JOIN ChatInference c
+  ON i.id_uint = toUInt128(c.id)
+LEFT JOIN JsonInference j
+  ON i.id_uint = toUInt128(j.id)
+WHERE uint_to_uuid(i.id_uint) = {id:String}
+FORMAT JSONEachRow;"#
+                .to_string(),
+            Some(&HashMap::from([("id", inference_id.to_string().as_str())])),
+        )
+        .await?;
+
+    if result.is_empty() {
+        return Err(Error::new(ErrorDetails::InvalidRequest {
+            message: format!("Inference `{}` not found", inference_id),
+        }));
+    }
+
+    let inference_data: TaggedInferenceDatabaseInsert =
+        serde_json::from_str(&result).map_err(|e| {
+            Error::new(ErrorDetails::Serialization {
+                message: format!("Failed to deserialize inference data: {}", e),
+            })
+        })?;
+    Ok(inference_data)
+}
+
+/// A handler for the feedback endpoint
```

Comment:
need to change this. 

Also give a better comment documenting behavior here please!